File: src/com/telmomenezes/synthetic/Evo.java
package com.telmomenezes.synthetic;

import java.io.BufferedWriter;
import java.io.FileWriter;

import com.telmomenezes.synthetic.io.NetFileType;
import com.telmomenezes.synthetic.random.RandomGenerator;


public class Evo {
    // best generators
	private Generator bestGenerator;
	private Generator bestFitGenerator;
	private double bestFitnessMax;
	private double bestFitnessAvg;
	
    // parameters
	private Net targNet;
	private int generations;
	private Generator baseGenerator;
	private String outDir;
	
    // state
	private int curgen;
	private double genTime;
	private double simTime;
	private double fitTime;
	
    private int bestCount;
    
    private MetricsBag targBag;
    
    private int bins;
    
    private double tolerance;
    
	
	public Evo(Net targNet, int generations, int bins, double tolerance, Generator baseGenerator, String outDir) {
		this.targNet = targNet;
		this.generations = generations;
		this.baseGenerator = baseGenerator;
		this.outDir = outDir;
		this.bins = bins;
		this.tolerance = tolerance;
		
		targBag = new MetricsBag(targNet, bins);
		writeDistribs(targBag, "targ");
	}
	
	
	public void run() {
		// init state
		genTime = 0;
		simTime = 0;
		fitTime = 0;
		bestCount = 0;
		writeLogHeader();
	
		// init population
		bestFitGenerator = baseGenerator.instance();
		bestFitGenerator.initRandom();
		bestFitGenerator.run();
		bestFitGenerator.computeFitness(targBag, bins);
		bestFitGenerator.getNet().clean();
		bestGenerator = bestFitGenerator;
		
		bestFitnessMax = bestFitGenerator.fitnessMax;
		bestFitnessAvg = bestFitGenerator.fitnessAvg;
		
		// evolve
		int stableGens = 0;
		for(curgen = 0; stableGens < generations; curgen++) {
			stableGens++;
			
			long startTime = System.currentTimeMillis();
			
			simTime = 0;
			fitTime = 0;
			
			Generator generator;
			if (RandomGenerator.random.nextBoolean()) {
				generator = bestFitGenerator.clone();
			}
			else {
				generator = bestGenerator.clone();
			}
			
			generator = generator.mutate();

			long time0 = System.currentTimeMillis();
			generator.run();
			simTime += System.currentTimeMillis() - time0;
			time0 = System.currentTimeMillis();
			generator.computeFitness(targBag, bins);
			generator.getNet().clean();
			fitTime += System.currentTimeMillis() - time0;	
			
			if (generator.isBetterThan(bestFitGenerator, bestFitnessMax, bestFitnessAvg, 0)) {
				bestFitnessMax = generator.fitnessMax;
				bestFitnessAvg = generator.fitnessAvg;
				bestFitGenerator = generator;
				stableGens = 0;
			}
			
			if (generator.isBetterThan(bestGenerator, bestFitnessMax, bestFitnessAvg, tolerance)) {
				bestGenerator = generator;
				onNewBest();
				stableGens = 0;
			}

			// time it took to compute the generation
			genTime = System.currentTimeMillis() - startTime;
			genTime /= 1000;
			simTime /= 1000;
			fitTime /= 1000;
			
			System.out.println("stable generation: " + stableGens);
			onGeneration();
		}
		
		System.out.println("Done.");
	}
    
	
    private void onNewBest() {
        String suffix = "" + bestCount + "_gen" + curgen;
        Generator bestGen = bestGenerator;
        
        // write net
        bestGen.getNet().save(outDir + "/bestnet" + suffix + ".txt", NetFileType.SNAP);
        bestGen.getNet().save(outDir + "/bestnet" + ".txt", NetFileType.SNAP);
        
        // write progs
        bestGen.getProg().write(outDir + "/bestprog" + suffix + ".txt");
        bestGen.getProg().write(outDir + "/bestprog" + ".txt");
        
        // write distribs
        writeDistribs(bestGen.getNet().metricsBag, "best");
        
        bestCount++;
    }

    private void writeLogHeader() {
    	// write header of evo.csv
    	try {
    		FileWriter fwriter = new FileWriter(outDir + "/evo.csv");
    		BufferedWriter writer = new BufferedWriter(fwriter);
    		if (isDirected()) {
    			writer.write("gen,best_fit_max,best_fit_avg,best_geno_size,gen_comp_time,sim_comp_time,fit_comp_time,in_degrees_dist,out_degrees_dist,d_pageranks_dist,u_pageranks_dist,triadic_profile_dist,d_dists_dist,u_dists_dist\n");
    		}
    		else {
    			writer.write("gen,best_fit_max,best_fit_avg,best_geno_size,gen_comp_time,sim_comp_time,fit_comp_time,degrees_dist,u_pageranks_dist,triadic_profile_dist,u_dists_dist\n");
    		}
    		writer.close() ;
    	}
    	catch (Exception e) {
    		e.printStackTrace();
    	}
    }
    
    
    private void writeDistribs(MetricsBag bag, String prefix) {
    	if (isDirected()) {
    		bag.getInDegrees().write(outDir + "/" + prefix + "_in_degrees.csv", false);
    		bag.getOutDegrees().write(outDir + "/" + prefix + "_out_degrees.csv", false);
    		bag.getDPageRanks().write(outDir + "/" + prefix + "_d_pagerank.csv", false);
    		bag.getdDists().write(outDir + "/" + prefix + "_d_dists.csv", false);
    	}
    	else {
    		bag.getDegrees().write(outDir + "/" + prefix + "_degrees.csv", false);
    	}
    	bag.getUPageRanks().write(outDir + "/" + prefix + "_u_pagerank.csv", false);
    	bag.getTriadicProfile().write(outDir + "/" + prefix + "_triadic_profile.csv", false);
    	bag.getuDists().write(outDir + "/" + prefix + "_u_dists.csv", false);
    }
    
    
    private void onGeneration() {
    	if (isDirected()) {
    		onGenerationDirected();
    	}
    	else {
    		onGenerationUndirected();
    	}
    }
    
    
    private void onGenerationDirected() {
        Generator bestGen = bestGenerator;
        double inDegreesDist = bestGen.getNet().metricsBag.getInDegreesDist();
        double outDegreesDist = bestGen.getNet().metricsBag.getOutDegreesDist();
        double dPageRanksDist = bestGen.getNet().metricsBag.getDPageRanksDist();
        double uPageRanksDist = bestGen.getNet().metricsBag.getUPageRanksDist();
        double triadicProfileDist = bestGen.getNet().metricsBag.getTriadicProfileDist();
        double dDistsDist = bestGen.getNet().metricsBag.getdDistsDist();
        double uDistsDist = bestGen.getNet().metricsBag.getuDistsDist();
        
        // write evo log
        try {
            FileWriter fwriter = new FileWriter(outDir + "/evo.csv", true);
            BufferedWriter writer = new BufferedWriter(fwriter);
            writer.write("" + curgen + ","
                    + bestFitnessMax + ","
                    + bestFitnessAvg + ","
                    + bestGenerator.getProg().size() + ","
                    + genTime + ","
                    + simTime + ","
                    + fitTime + ","
                    + inDegreesDist + ","
                    + outDegreesDist + ","
                    + dPageRanksDist + ","
                    + uPageRanksDist + ","
                    + dDistsDist + ","
            		+ uDistsDist + ","
                    + triadicProfileDist + "\n");
            writer.close();
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        
        System.out.println(genInfoString());
        System.out.println("inDegreesDist: " + inDegreesDist + "; outDegreesDist: " + outDegreesDist + "; dPageRanksDist: " + dPageRanksDist + "; uPageRanksDist: " + uPageRanksDist + "; triadicProfileDist: " + triadicProfileDist + "; dDistsDist: " + dDistsDist + "; uDistsDist: " + uDistsDist);
    }
	
    
    private void onGenerationUndirected() {
        Generator bestGen = bestGenerator;
        double degreesDist = bestGen.getNet().metricsBag.getDegreesDist();
        double uPageRanksDist = bestGen.getNet().metricsBag.getUPageRanksDist();
        double triadicProfileDist = bestGen.getNet().metricsBag.getTriadicProfileDist();
        double uDistsDist = bestGen.getNet().metricsBag.getuDistsDist();
        
        // write evo log
        try {
            FileWriter fwriter = new FileWriter(outDir + "/evo.csv", true);
            BufferedWriter writer = new BufferedWriter(fwriter);
            writer.write("" + curgen + ","
                    + bestFitnessMax + ","
                    + bestFitnessAvg + ","
                    + bestGenerator.getProg().size() + ","
                    + genTime + ","
                    + simTime + ","
                    + fitTime + ","
                    + degreesDist + ","
                    + uPageRanksDist + ","
            		+ uDistsDist + ","
                    + triadicProfileDist + "\n");
            writer.close();
        }
        catch (Exception e) {
            e.printStackTrace();
        }
        
        System.out.println(genInfoString());
        System.out.println("degreesDist: " + degreesDist + "; uPageRanksDist: " + uPageRanksDist + "; triadicProfileDist: " + triadicProfileDist + "; uDistsDist: " + uDistsDist);
    }
    
    
    public boolean isDirected() {
    	return targNet.directed;
    }
    
	
	public String infoString()
	{
		String str = "stable generations: " + generations + "\n";
		str += "directed: " + isDirected() + "\n";
		str += "target net node count: " + targNet.getNodeCount() + "\n";
        str += "target net edge count: " + targNet.getEdgeCount() + "\n";
        str += "distribution bins: " + bins + "\n";
        str += "tolerance: " + tolerance + "\n";
		return str;
	}


	private String genInfoString()
	{
		String tmpstr = "gen #" + curgen
        	+ "; best fitness max: " + bestFitnessMax
        	+ "; best fitness avg: " + bestFitnessAvg
        	+ "; best genotype size: " + bestGenerator.getProg().size()
        	+ "; gen comp time: " + genTime + "s."
			+ "; sim comp time: " + simTime + "s."
			+ "; fit comp time: " + fitTime + "s.";
        	return tmpstr;
	}
}

File: netgens/evo.py
import random
from netgens.utils import *


def within_tolerance(fitness, best_fitness, tolerance):
    return abs(fitness - best_fitness) < tolerance


class EvaluatedIndividual(object):
    def __init__(self, evo, generator, net):
        self.generator = generator
        self.fitness_max, self.fitness_avg, self.fitness_tuple = evo.compute_fitness(net)

    def is_better_than(self, eval_indiv, best_fitness, tolerance):
        fitness_orig = self.fitness_max
        fitness_targ = eval_indiv.fitness_max

        if tolerance <= 0:
            return fitness_orig < fitness_targ

        if within_tolerance(fitness_orig, best_fitness, tolerance):
            if not within_tolerance(fitness_targ, best_fitness, tolerance):
                return True
            else:
                return self.generator.prog.size() < eval_indiv.generator.prog.size()
        return False


class Evo(object):
    def __init__(self, distances, generations, tolerance, base_generator, out_dir):
        self.distances = distances
        self.generations = generations
        self.base_generator = base_generator
        self.out_dir = out_dir
        self.tolerance = tolerance

        # best individuals
        self.best_individual = None
        self.best_fit_individual = None

        # state
        self.curgen = 0
        self.best_count = 0

        # timers
        self.gen_time = 0
        self.sim_time = 0.
        self.fit_time = 0.

    def compute_fitness(self, net):
        dists = self.distances.compute(net)
        fitness_max = max(dists)
        fitness_avg = sum(dists) / len(dists)
        return fitness_max, fitness_avg, dists

    def run(self):
        # init state
        self.gen_time = 0
        self.sim_time = 0
        self.fit_time = 0
        self.best_count = 0
        self.write_log_header()

        # init population
        generator = self.base_generator.spawn_random()
        net = generator.run()
        self.best_fit_individual = EvaluatedIndividual(self, generator, net)
        self.best_individual = self.best_fit_individual

        # evolve
        stable_gens = 0
        self.curgen = 0
        while stable_gens < self.generations:
            self.curgen += 1
            stable_gens += 1

            start_time = current_time_millis()

            sim_time = 0
            fit_time = 0

            if random.choice([True, False]):
                generator = self.best_fit_individual.generator.clone()
            else:
                generator = self.best_individual.generator.clone()

            generator = generator.mutate()

            time0 = current_time_millis()
            net = generator.run()
            sim_time += current_time_millis() - time0
            time0 = current_time_millis()
            individual = EvaluatedIndividual(self, generator, net)
            fit_time += current_time_millis() - time0

            best_fitness_max = self.best_fit_individual.fitness_max
            if individual.is_better_than(self.best_fit_individual, best_fitness_max, 0):
                self.best_fit_individual = individual
                stable_gens = 0

            if individual.is_better_than(self.best_individual, best_fitness_max, self.tolerance):
                self.best_individual = individual
                self.on_new_best()
                stable_gens = 0

            # time it took to compute the generation
            gen_time = current_time_millis() - start_time
            gen_time /= 1000
            sim_time /= 1000
            fit_time /= 1000

            print('stable generation: %s' % stable_gens)
            self.on_generation()

        print('Done.')

    def on_new_best(self):
        suffix = '%s_gen%s' % (self.best_count, self.curgen)
        best_gen = self.best_individual.generator
        
        # write net
        best_gen.net.save('%s/bestnet%s.gml' % (self.out_dir, suffix))
        best_gen.net.save('%s/bestnet.gml' % self.out_dir)
        
        # write progs
        best_gen.prog.write('%s/bestprog%s.txt' % (self.out_dir, suffix))
        best_gen.prog.write('%s/bestprog.txt' % self.out_dir)
        
        self.best_count += 1

    def write_log_header(self):
        # write header of log file
        with open('%s/evo.csv' % self.out_dir, 'w') as log_file:
            header = 'gen,best_fit_max,best_fit_avg,best_geno_size,gen_comp_time,sim_comp_time,fit_comp_time'
            stat_names = [stat.name() for stat in self.distances.targ_stats_set.stats]
            header = '%s,%s\n' % (header, ','.join(stat_names))
            log_file.write(header)

    def on_generation(self):
        dists = [str(dist) for dist in self.best_individual.fitness_tuple]

        # write log line for generation
        with open('%s/evo.csv' % self.out_dir, 'a') as log_file:
            row = ','.join((self.curgen, self.best_individual.fitness_max, self.best_individual.fitness_avg,
                            self.best_individual.prog.size(), self.gen_time, self.sim_time, self.fit_time))
            row = '%s,%s\n' % (row, ','.join(dists))
            log_file.write(row)

        # print info
        print(self.gen_info_string())
        stat_names = [stat.name() for stat in self.distances.targ_stats_set.stats]
        items = ['%s: %s' % (stat_names[i], dists[i]) for i in range(len(stat_names))]
        print('; '.join(items))

    def info_string(self):
        lines = ['stable generations: %s' % self.generations,
                 'directed: %s' % self.distances.net.is_directed(),
                 'target net node count: %s' % len(self.distances.net.vs),
                 'target net edge count: %s' % len(self.distances.net.es),
                 'distribution bins: %s' % self.distances.bins,
                 'tolerance: %s' % self.tolerance]
        return '\n'.join(lines)

    def gen_info_string(self):
        items = ['gen #%s' % self.curgen,
                 'best fitness max: %s' % self.best_individual.fitness_max,
                 'best fitness avg: %s' % self.best_individual.fitness_avg,
                 'best genotype size: %s' % self.best_individual.prog.size(),
                 'gen comp time: %ss.' % self.gen_time,
                 'sim comp time: %ss.' % self.sim_time,
                 'fit comp time: %ss.' % self.fit_time]
        return '; '.join(items)
